src/uu/df/src/table.rs

```diff
@@ -240,82 +238,161 @@ impl<'a> DisplayRow<'a> {
-                }
-                Column::Target => write!(f, "{0: <16}", self.row.fs_mount)?,
-                Column::Itotal => write!(f, "{0: >12} ", self.scaled(self.row.inodes)?)?,
-                Column::Iused => write!(f, "{0: >12} ", self.scaled(self.row.inodes_used)?)?,
-                Column::Iavail => write!(f, "{0: >12} ", self.scaled(self.row.inodes_free)?)?,
-                Column::Ipcent => {
-                    write!(f, "{0: >5} ", DisplayRow::percentage(self.row.inodes_usage))?;
-                }
-                Column::File => {
-                    write!(f, "{0: <16}", self.row.file.as_ref().unwrap_or(&"-".into()))?;
-                }
-                Column::Fstype => write!(f, "{0: <5} ", self.row.fs_type)?,
+            let string = match column {
+                Column::Source => self.row.fs_device.to_string(),
+                Column::Size => self.scaled(self.row.bytes),
+                Column::Used => self.scaled(self.row.bytes_used),
+                Column::Avail => self.scaled(self.row.bytes_avail),
+                Column::Pcent => Self::percentage(self.row.bytes_usage),
+
+                Column::Target => self.row.fs_mount.to_string(),
+                Column::Itotal => self.scaled(self.row.inodes),
+                Column::Iused => self.scaled(self.row.inodes_used),
+                Column::Iavail => self.scaled(self.row.inodes_free),
+                Column::Ipcent => Self::percentage(self.row.inodes_usage),
+                Column::File => self.row.file.as_ref().unwrap_or(&"-".into()).to_string(),
+
+                Column::Fstype => self.row.fs_type.to_string(),
                 #[cfg(target_os = "macos")]
-                Column::Capacity => write!(
-                    f,
-                    "{0: >12} ",
-                    DisplayRow::percentage(self.row.bytes_capacity)
-                )?,
-            }
+                Column::Capacity => Self::percentage(self.row.bytes_capacity),
+            };
+
+            strings.push(string);
         }
-        Ok(())
+
+        strings
     }
 }
 
-/// The header row.
-///
-/// The `options` control which columns are displayed.
-pub(crate) struct Header<'a> {
-    /// Options that control which columns are displayed.
-    options: &'a Options,
+/// The data of the header row.
+struct Header {}
+
+impl Header {
+    /// Return the headers for the specified columns.
+    ///
+    /// The `options` control which column headers are returned.
+    fn get_headers(options: &Options) -> Vec<String> {
+        let mut headers = Vec::new();
+
+        for column in &options.columns {
+            let header = match column {
+                // We use a header's width as the minimum column width. To make this approach work
+                // with the "Filesystem" column with its wider minimum column width, we have to pad
+                // its header.
+                Column::Source => String::from("Filesystem    "),
```

Comment:
Instead of trying to encode the minimum width in the length of the string, could we instead make an associated function on `Column`, say `Column::min_width()`? Then below in `Table::new()` you could have
```rust
         let mut widths: Vec<_> = options.columns.iter().map(Column::min_width).collect();
```

Reply:
Good point. I will move the minimum width to `Column`.